[WIP] Fix 3d shadowcasting

ifreund opened the pull request:
#### Summary
SUMMARY: Bugfixes "Fix bugs and increase performance of 3D shadowcasting"

#### Purpose of change
I'm attempting to revive #30861 as Kevin is busy with other performance work.  Much of this is directly copied from that PR.

Fixes #6821
BOUNTY

3D shadowcasting has a number of rendering bugs remaining, fix them.
A non-exhaustive list:

- ~~As in #29175 shockers are sometimes able to see and target you through a floor. With some debugging Kevin determined the cause to be 3D shadowcasting.~~ Was caused by map shifts apparently.
- When standing cardinally adjacent to a single opaque tile (wall or adult tree, etc), in 2D shadowcasting the area behind the tree is obscured, but in 3D it is visible.
UPDATE: seems to be fixed with the `slope` struct.
- When standing adjacent to a wall corner, in 2D shadowcasting the first tile of the wall is visible and the rest is occluded, in 3D the entire wall and anything behind it is visible.
UPDATE: seems to be fixed with the `slope` struct.
- 3D shadowcasting does not cast directly up or down, so there's a shadow directly above and below the player.
UPDATE: Now fixed.  We cast on 8 new octants, 4 up and 4 down.

#### Describe the solution
This is going to be a progressive process since I don't know what the problem is.
- [x] Extend shadowcasting test to trigger 3D shadowcasting and test 3D scenarios.
- [x] Add tests that characterize the breakage.
- [x] Replace recursive shadowcasting with iterative shadowcasting as outlined in #23373. In addition to being more performant, this should be easier to analyze and test.
- [x] Fix an infinite loop that pops up in some cases.
- [x] Fix the remaining bugs.
- [x] Cast straight up and down

kevingranade committed:
Extend shadowcasting test to 3D

kevingranade committed:
Add some 3D tests.

kevingranade committed:
Extract cast_zlight to a dedicated code module

kevingranade committed:
WIP de-recursify shadowcasting

ifreund committed:
Remove unnecessary floats and fix bugs

Replace floats with new zlight_slope struct.

This now passes 2 of the tests that the old recursive algorithm was failing,
but encounters an infinite loop in some cases without the fuzzing that seems
to have been hiding a major bug.

Fix a few other minor bugs.

ifreund committed:
More bugfixes and kill the inifinte loop

This is the best way I could find to fix the infinite loop without
creating tons of artifacts.

ifreund commented:
Quick update, I've corrected most of the major bugs and the new iterative 3d FOV is now totally playable and quite a bit more correct that the current recursive algorithm.

The only major todo item I'm currently aware of is doing something about casting straight up/down as we currently don't do that at all (That's the only test failing on travis currently).

ifreund committed:
Add more comments and clean up formatting

- Add casting direction diagrams
- Remove unnecessary declaration

ifreund committed:
Cast zlight straight up and down

This commit adds a new cast_vertical_zlight_segment function with
different iteration logic than the now renamed
cast_horizontal_zlight_segment.

This duplicates a lot of code, but imo is worth it for the readablity
gains over adding a lot more contitonals to the iteration logic to handle
zx and zy template parameters properly.

ifreund commented on src/shadowcasting.cpp:
```diff
@@ -0,0 +1,707 @@
+         T( *calc )( const T &, const T &, const int & ),
+         bool( *check )( const T &, const T & ),
+         T( *accumulate )( const T &, const T &, const int & )>
+void cast_vertical_zlight_segment(
```
Splitting `cast_zlight_segment` into vertical and horizontal versions isn't ideal and duplicates a lot of complex code.  However, the alternative is some quite nasty iteration logic to handle the full range of possible coordinate transformations instead of the subset currently handled.  Splitting it like this makes this already complex code a good bit more readable without the extra conditonals.

Would definitely appreciate feedback on whether this choice makes sense to others though.

jbytheway replied:
It's certainly possible to merge.  You'd want to abstract out the coordinate system more into some type template argument that would replace the various `_transform` numbers.  Each version of that type would have methods to do the various transformations needed (between depth/major/minor coordinates and x/y/z coordinates).

I think it is worth doing and would probably help clarity and maintainability, but it doesn't have to be on this first PR.

ifreund commented:
3d shadowcasting now casts directly up and down, with a total of 8 more casts than the previous method.  This means no more pyramid of unseen tiles directly above and below the player.

Performance impact shouldn't be too bad, as downward casts will almost always end immediately as they encounter the opaque ground and upward casts will either end immediately if under a roof or end much sooner than horizontal casts due to our rather low maximum z level.

So, upwards casting will be a slight slowdown when outdoors, but allows the player to see all the tiles they are supposed to see.  The optimization described in #23914 or something like it would make upward casting trivial in most cases as the area above the player is usually either blocked by a roof or completely open.

ifreund committed:
Add a helper to reduce code duplication

kevingranade commented on src/shadowcasting.cpp:
```diff
@@ -0,0 +1,659 @@
+        }
+    }
+
+    int rise;
```
Fun fact, the shadowcasting area is only 60 deep and 60 wide, so rise and run can be 8-bit.
probably int_least8_t or uint_least8_t

ifreund replied:
Makes sense, I had meant to go back and revise that but forgot.

ifreund replied:
We do have negative slopes in some valid cases, so I'll be using `int_least8_t`

jbytheway replied:
FWIW, I think `int_least8_t` is probably overkill.  Cataclysm is already not going to work on any system where that isn't the same as `int8_t`.

ifreund committed:
Use int_least8_t for rise and run

ifreund commented:
Need to figure out if `int_fast8_t` would be better for this use case, but that is non-trivial to profile properly and is an small change if needed.

Travis is complaining about a memory leak and I'm not quite sure how that could have happened as a result of these changes, but I'll do some more digging.

kevingranade commented on tests/shadowcasting_test.cpp:
```diff
@@ -412,197 +413,406 @@ static void shadowcasting_3d_2d( const int iterations )
+        trans_grid << '\n';
+        expected_grid << '\n';
+        actual_grid << '\n';
+        delete caches[gz];
```
Here's your leak, it allocates either one or 21 cache levels depending on FoV flag, but only deletes the ones that have a corresponding test case entry.

ifreund replied:
Perfect thanks, for some reason my leak profiler was showing some memory leakage on loading a normal game so I wasn’t looking at the tests.

ifreund replied:
And yes I have realized now that asan totally gave the function name for where the leak was, I was running on very little sleep when I first looked at this.

ifreund committed:
Fix memory leak

ifreund committed:
Cleanup template parameters

ifreund commented:
Clany tidy error is unrelated, I'd say this is good to go.

ifreund commented:
Jenkins rebuild

ZhilkinSerg commented:
Jenkins rebuild.

esotericist commented:
So, I merged this locally to do some playtests, and there's some kinda weird behavior here with view distance in reduced vision situations. I first noticed it while in a vehicle, but I was able to reproduce it in a building.

Details: it's 5:16 AM, with "drizzle" weather. So it's "very dark" outside.

Default vision (3d fov off) while in vehicle:
![default vision in vehicle](https://cdn.discordapp.com/attachments/511623330227421184/597616608944455681/unknown.png)

After turning on 3d fov, and moving (to reset visible tiles):
![3d fov in vehicle](https://cdn.discordapp.com/attachments/511623330227421184/597616659976421377/unknown.png)

Here's the difference between not-3d-fov and yes-3d-fov in the building (also involving steps in between to reset visible tiles):

![default vision in building](https://cdn.discordapp.com/attachments/511623330227421184/597617545130016768/unknown.png)

![3d fov in building](https://cdn.discordapp.com/attachments/511623330227421184/597617570056503337/unknown.png)

For extra bonus weird, I noticed view distance changing as I moved within the building. These next two shots are still with 3d fov on, as I step towards the door:

![approaching door](https://cdn.discordapp.com/attachments/511623330227421184/597617785828540416/unknown.png)

![approaching door more](https://cdn.discordapp.com/attachments/511623330227421184/597617819156217857/unknown.png)

The behavior reverses as I move back to my original position.

Extra extra extra bonus weird: what happens if I move one tile to the right, to sit in the chair:

![image](https://user-images.githubusercontent.com/1569754/60779532-f3d59700-a0ef-11e9-8f3a-be1af30a893e.png)

That last one at least matches the default vision algorithm

jbytheway commented:
I had to [turn off one of the vision tests](https://github.com/CleverRaven/Cataclysm-DDA/blob/9d36f0db4f03603bc8d30bc02851fc89a69710b0/tests/vision_test.cpp#L373) on 3D because of the inconsistencies.  Please turn that back on (indeed, you can totally rip out the feature of being able to turn off particular tests in 3D mode).

jbytheway commented on src/shadowcasting.cpp:
```diff
@@ -0,0 +1,661 @@
+ * Handle splitting the current span in cast_horizontal_zlight_segment and
+ * cast_vertical_zlight_segment to avoid as much code duplication as possible
+ */
+template<typename T, bool( *check )( const T &, const T & ), T( *accumulate )( const T &, const T &, const int & )>
```
Would perhaps be clearer if `check` was called something like `is_transparent`?

ifreund replied:
Will fix

jbytheway reviewed (commented):
Looking good in terms of code quality.  I haven't tested it, and the above examples indicate some changes from the previous algorithm, which should be investigated.

I'd like to see the vision tests generalized to 3D as well as the shadowcasting tests.  The shadowcasting tests only test the boolean "is it visible at all", whereas the vision tests get into questions like "how visible is it?".  I'd like us to be able to verify that there aren't any artefacts at the boundaries between the horizontal and vertical casting algorithms (e.g. look three levels up, is there an obvious 6x6 square where stuff looks different).  Not vital that those be added in this PR, though.

ifreund commented on src/shadowcasting.cpp:
```diff
@@ -0,0 +1,661 @@
+}
+inline bool operator==( const slope &lhs, const slope &rhs )
+{
+    // a/b == c/d <=> a*d == c*b if b and d have the same sign.
```
Of course, it's the inequality operators where it matters

ifreund committed:
Test 3d fov in all vision tests

and remove the code that allows disabling it

ifreund committed:
Rename check to is_transparent and fix comments

It somehow took me until now to realize that y coordinates are counted
from the "top" of the screen not the bottom, so fix the comment
diagrams

ifreund committed:
Fix erroneous cumulative transparency incrementation

ifreund commented:
I believe my most recent commit has fixed that bug @esotericist, at least I can't reproduce it anymore.  Thanks for testing!

jbytheway commented:
> I believe my most recent commit has fixed that bug @esotericist, at least I can't reproduce it anymore. Thanks for testing!

Did you ensure we have a test that captures that failure mode before fixing it?

jbytheway commented on tests/vision_test.cpp:
```diff
@@ -284,18 +283,13 @@ struct vision_test_case {
     void test_all() const {
-        // Disabling 3d tests for now since 3d sight casting is actually
-        // different (it sees round corners more).
-        if( test_3d ) {
```
The `if` can go, but the braces need to stay, so that only one `INFO` from this function is in scope for each test (that affects which show up in Catch2 failure reports).

ifreund replied:
Fixed

ifreund commented:
>Did you ensure we have a test that captures that failure mode before fixing it?

Unfortunately I haven't been able to setup a test case that consistently passes after the change and fails before it, since the failure case is very sensitive to lighting.  I've tried setting the weather before the test and doing the test several z levels underground.  I can get it to behave properly 4/5 times or so, but haven't managed to track down the inconsistency yet.

kevingranade commented:
This looks not-right...
![2019-07-09-070254_1366x768_scrot](https://user-images.githubusercontent.com/860276/60866414-d8fe4200-a217-11e9-9b16-073ac9e7299d.png)
![2019-07-09-070300_1366x768_scrot](https://user-images.githubusercontent.com/860276/60866415-d996d880-a217-11e9-9ac1-84264d08c3cc.png)
![2019-07-09-070303_1366x768_scrot](https://user-images.githubusercontent.com/860276/60866416-d996d880-a217-11e9-87fa-ccaa834e628b.png)
![2019-07-09-070311_1366x768_scrot](https://user-images.githubusercontent.com/860276/60866417-d996d880-a217-11e9-9075-1d43b293428d.png)
![2019-07-09-070314_1366x768_scrot](https://user-images.githubusercontent.com/860276/60866418-d996d880-a217-11e9-945f-96c7e4dc6124.png)
![2019-07-09-070317_1366x768_scrot](https://user-images.githubusercontent.com/860276/60866419-d996d880-a217-11e9-8897-d4c26f98ef95.png)

ifreund commented:
Went to the same location and could only reproduce half of the bug.  You sure you had 96abcadc99a4d972e3cef597a52383bd5b75d75c included in that build?
<details>
  <summary>Screenshots</summary>
   
![Screen Shot 2019-07-09 at 10 06 06 AM](https://user-images.githubusercontent.com/12723818/60870713-ad885100-a231-11e9-9f1e-327711c3027c.png)
![Screen Shot 2019-07-09 at 10 06 18 AM](https://user-images.githubusercontent.com/12723818/60870726-b1b46e80-a231-11e9-8a65-c041cbbc1c19.png)
![Screen Shot 2019-07-09 at 10 06 25 AM](https://user-images.githubusercontent.com/12723818/60870743-b6792280-a231-11e9-9aee-baf9290824f0.png)
![Screen Shot 2019-07-09 at 10 06 30 AM](https://user-images.githubusercontent.com/12723818/60870756-baa54000-a231-11e9-824b-d2789ecdb080.png)


</details>

ifreund commented:
Marking this WIP for now as there's still a reproducible bug with it that I haven't managed to track down yet.  In the process of debugging I found some faulty logic that may be partially at fault and am in the process of reworking how spans get split.

ifreund committed:
WIP rewrite of span splitting

ifreund committed:
Fix some of the bugs

ThomasLinkin commented:
so this is abandoned i suppose?

esotericist commented:
> so this is abandoned i suppose?

Not abandoned, merely stalled. The people with the appropriate expertise have been dealing with other matters, be it within the project or things outside of the project.

This *will* get tended to eventually.

ifreund commented:
Superseded by 42835

Brambor commented:
> Superseded by #42835

Add # so that I can click it :)
